### DISCUSSION TITLE
ZipChord 0: A roadmap to faster, cross-platform ZipChord

### DISCUSSION THREAD

**Discussion by psoukie** [psoukie/zipchord#203] [Announcements]
The last redesign (as described in #145) brought the ability to chain chords, but more importantly, the restructuring and refactoring of the code has made implementing recent (and future) ZipChord improvements much easier.

Still, ZipChord has inherent limitations: While the AutoHotkey programming language enabled me to take the plunge to build the first tool (to my knowledge) that could reliably combine chords of more than three keys with regular typing, it has drawbacks: the language is error-prone, it limits ZipChord to Windows only, and although ZipChord runs well even with large dictionaries, more performance is always better.

My original idea for where ZipChord should go next was to port the code to AutoHotkey&nbsp;v2. But this would be extremely impractical because there is no way to port parts of code, and it would not improve my areas of concern. It led me to thinking about a solution that could be implemented gradually _and_ would addresses the platform, error-checking while programming, and performance limitations: 

## ZipChord Z

<img align="right" src="https://github.com/psoukie/zipchord/assets/42703196/c3191ae1-44a8-4c8f-9bb3-97b8d594cc70" width=400/> 

First, what's in the name? I chose the name ZipChord&nbsp;Z for three reasons:[^fn]
- Z alliterates with Zipf's law (which ZipChord is named after),
- It is associated with speed (at least to my ears; and I think Nissan thought so too),
- And it stands for [Zig](https://github.com/ziglang/zig), the programming language behind the gradual transformation.

ZipChord&nbsp;Z is a design concept rather than a specific release. Future versions will start including alternative implementations of specific modules as dynamic libraries written in Zig. (The first one will be the dictionary code responsible for loading and looking up chords and shorthands.)

New releases will continue to function with AutoHotkey-only code (whether as executable or uncompiled AHK script), but if the new Zig-compiled dynamic library will be detected, available functions will be directed to a faster—and, crucially, cross-platform-capable—Zig implementation.

When enough portions of the code have their Zig alternatives, it will make sense to stop maintaining the AutoHotkey versions in parallel. At this point, the Zig libraries will start being required, and the work will continue on porting the remaining platform-independent parts of code into Zig.

Obviously, some modules (including keyboard input and output modifications) need to have different implementations for each platform. These will be converted last. At that point, ZipChord will be fully cross-platform, albeit without graphical user interface. Native user interfaces will be the last step and are likely going to use a third-party Zig cross-platform library for user interfaces.

My goal is to continue working on ZipChord at a sustainable pace. Don't hesitate to continue submitting suggestions and feature requests. The nice part of the ZipChord Z concept is that it can be done in parallel with other improvements to ZipChord, which was not something that was possible with the first redesign.

I have already created a small proof-of-concept implementation, and I look forward to sharing the first ZipChord&nbsp;Z release soon.

**Comment by jakub-wojciech** [comment]
I would be really happy to look at it. Do you have any progress with it?

**Comment by Mammal4963** [comment]
I noticed that hints seemed to have disappeared from this diagram. Are the on screen display hints for chords still a feature that will remain a part of Zipchord Z?

**Comment by psoukie** [comment]
Yes, for sure, the hints will remain. They are under the User Interface block (at least in my mind), and will require a platform-specific implementation. There was a long delay in my ability to develop, but looking forward to make some progress over December.

**Reply by Mammal4963** [reply]
Happy to hear that, you have earned at least 3 new users, myself and 2 others in my office have been using your program heavily for about a month and a half. I personally have benefitted so much, it makes typing such a better experience. I manage our chord dictionary and as of writing this we have added at least 340 words to the dictionary. Only sharing this because you definitely deserve to hear the praise, for what that is worth.

**Comment by psoukie** [comment]
A long-overdue update on the roadmap:

Since I created this plan, I took two long pauses in the work on this project. There was also one other critical thing I really wanted to implement first: Make ZipChord map to specific physical keys rather than to the characters that are produced by those keys. The distinction matters only when dealing with keyboard layouts in different languages, but it is fundamental.

AutoHotkey's logic is to identify keys by the characters they produce. So the physical key located where `Y` is on QWERTY layout is conveniently treated as multiple keys (Y, Z, etc.) depending on the active keyboard layout. This is easy and straightforward, until it isn't. Specifically, in languages with extended characters (Czech, Polish, ...), there is no nice way to make the former approach play along.

I had a version of ZipChord that overhauled keyboard detection to physical keys (including automated detection of the key layout) working and sitting in the repository since October 2025. But I did not have a good way to deal with application keyboard shortcuts and special keys for non-printing characters in a way that's compatible with the new approach. Then I started a new job and several months went by.

The pause was useful. It allowed me to get enough distance to see there were parts of ZipChord with tangential, obscure functionality that were just much better thrown away (despite the time and effort spent on coding them). It was suddenly much clearer to see that allowing mapping of keys like PrintScreen and F-keys (with an optional bypass of original function on top) was just not in line with the goal. It was also easy to see how a command menu is a much better alternative to the customizable application keyboard shortcuts with short or long presses of hotkeys such as Ctrl-X.

With version 2.6 having taken care of the keyboard layout issue (and given my career-long involvement in localization, this was embarrassingly overdue), it was time to revisit the plan outlined two years ago in this post.

At the time of the original post, I created a minimal proof of concept to write and compile a Zig library that could be leveraged by AutoHotkey code. Obviously, Zig can do things like that; the question was if I could.

With hindsight, I realized that even though Zig is a language I wanted to love, it is a language that makes programming feel like a chore and does not come naturally to me.

I also knew of Odin language for many years but didn't think it was as "worthy" to learn. But then I realized a few things: I spent time back then to go through the Ziglings exercises to learn Zig and coded the proof of concept, and yet, I was much more able to write in Odin after just reading parts of the book on Odin by Karl Zylinski without any practice.

This is not to say "Odin is better." Zig and Odin have different philosophies and both are worthy, serious languages. I happened to grow up with an 8-bit Atari on Basic and switched to Pascal before learning other programming languages. While switching to a lower level language is a huge adjustment, there's just something about Odin (and maybe Pascal's influence on Odin is a part of it) that makes writing it instinctively make sense to me.

For a project that's done in my free time, having a tool that doesn't feel like it wants to punish me because I need to manipulate strings in UTF-8 and that "speaks the same language" is huge. It's the difference between something feeling tedious and having a smile on my face when I see the magic happen.

Does this change anything about my plan above? Not really, other than switching the code name from ZipChord Z to ZipChord 0 (with the slashed "zero" standing for Odin) and the programming language to one that is easier and faster to use for me while offering the same cross-platform and performance benefits.

I can safely say this is the first time I'm truly having fun writing ZipChord. (This includes all the AutoHotkey code too: it was the only tool I knew when I started that supported the keyboard handling I needed.) I currently have a parallel compiled dynamic library (written in Odin) that works as an on-the-fly replacement for the same dictionary functions of ZipChord (loading and looking up things in dictionaries). Its significance is now mostly symbolic. Yes, it loads dictionaries 10x faster, but the lookups are only modestly faster because of the time consumed by DLL calls. But that's not the main point.

The real point is, it is the first truly tangible step on the way to cross-platform ZipChord. And yes, as the compiled library grows to include alternatives for the other parts of ZipChord's keyboard input and module processing (which are the bigger performance bottleneck), the performance gains will be meaningful even during typing. But the really important thing is that Odin opens ZipChord's future for other platforms and does that in a way that makes me want to work on it when I can.

**Reply by Mammal4963** [reply]
Good for you! A lot of people don't understand what it takes to maintain a project in their free time, or that there is often a gap between the things we want to achieve and the knowledge we have to achieve with. Often we need time to grow into our aspirations, most of the time (at least for me personally) I just know that I don't know enough yet and am on the lookout for that knowledge nugget or that missing piece that my project needs. 

Just wanted to validate your post, by saying that I get it. It sounds like you know where your project needs to go and that you'll be the one to take it there. We will all be here for it.
